searx/engines/mullvad_leta.py

```diff
@@ -0,0 +1,200 @@
+
+    country = get_country(traits)
+    params['url'] = search_url
+    params['method'] = 'POST'
+    params['data'] = {
+        "q": query,
+        "gl": country,
+    }
+    # pylint: disable=undefined-variable
+    if use_cache:
+        params['data']['oc'] = "on"
+    # pylint: enable=undefined-variable
+
+    if params['time_range'] in time_range_dict:
+        params['dateRestrict'] = time_range_dict[params['time_range']]
+    else:
+        params['dateRestrict'] = ''
+
+    if params['pageno'] > 1:
+        #  Page 1 is n/a, Page 2 is 11, page 3 is 21, ...
+        params['data']['start'] = ''.join([str(params['pageno'] - 1), "1"])
+
+    if params['headers'] is None:
+        params['headers'] = {}
+
+    assign_headers(params['headers'])
+    return params
+
+
+def response(resp: Response):
+    """post-response callback
+    resp: requests response object
+    """
+
+    def extract_result(dom_result: html.HtmlElement):
+        [a_elem, h3_elem, p_elem] = eval_xpath_list(dom_result, 'div/div/*')
+        return {
+            'url': extract_text(a_elem.text),
+            'title': extract_text(h3_elem),
+            'content': extract_text(p_elem),
+        }
+
+    # def are_more_pages(dom: html.HtmlElement) -> bool:
+    #    previous and next buttons. If only 1, then just previous
+    #    return len(eval_xpath_list(dom, '//main/div[3]/form')) == 2
+    # def get_page_x_of_y(dom: html.HtmlElement) -> Tuple[int, int]:
+    #    page_x_of_y = extract_text(eval_xpath(dom.body, '//main/div[3]/span'))
+    #    if page_x_of_y is None or page_x_of_y == '':
+    #        return (0, 0)
+    #    [_, x, _, y] = page_x_of_y.split(' ')
+    #    return (int(x), int(y))
+    dom = html.fromstring(resp.text)
+    if not is_vpn_connected(dom):
+        raise SearxEngineResponseException('Not connected to Mullvad VPN')
+    search_results = eval_xpath(dom.body, '//main/div[2]/div')
+    results: List[dict] = [extract_result(sr) for sr in search_results]
+    # (x, y) = get_page_x_of_y(dom)
+    # results.append({
+    #    'number_of_results': len(search_results),
+    #    'pages': y-x+1
+    # })
+    return results
+
+
+def fetch_traits(engine_traits: EngineTraits):
+    """Fetch languages and regions from Mullvad-Leta"""
+    # pylint: disable=import-outside-toplevel
+    # see https://github.com/searxng/searxng/issues/762
+    from searx.network import post as http_post
+
+    # pylint: enable=import-outside-toplevel
+    resp = http_post(search_url, headers=assign_headers({}))
+    if not isinstance(resp, Response):
+        print("ERROR: failed to get response from mullvad-leta. Are you connected to the VPN?")
+        return
+    if not resp.ok:
+        print("ERROR: response from mullvad-leta is not OK. Are you connected to the VPN?")
+        return
+    dom = html.fromstring(resp.text)
+    if not is_vpn_connected(dom):
+        print('ERROR: Not connected to Mullvad VPN')
+        return
```

Comment:
This function will be called by gh worklfows when [updating data](https://github.com/searxng/searxng/blob/master/.github/workflows/data-update.yml) .. when we are not in the Mullvad VPN.

Reply:
Will this even run if Mullvad is commented out in settings? It technically works, so I don't see a reason not to include it.

Reply:
Sorry / forgot that engine is commented out by default .. that means:

**Maintainers have to run `make data.traits` (in the Mullvad VPN) and merge the modified JSON file `searx/data/engine_traits.json` ... on every single update of SearXNG?** /  May this is not clear to maintainers and needs to be documented?

Reply:
> Maintainers have to run make data.traits (in the Mullvad VPN) and merge the modified JSON file searx/data/engine_traits.json ... on every single update of SearXNG?

Wasn't aware the maintainers would need to run the update on every update. Let's keep it simple, I'll just remove this method and have no engine traits.

Reply:
I've included a dedicated WARNING header for this specific issue.

I feel like its a waste to delete, but it likely won't ever be ran outside of local usage (unless Mullvad wants to sponsor us with a free VPN connection :smile: )

Reply:
Although now merged: Can you use the logger instead of print statements?

Reply:
> Although now merged: Can you use the logger instead of print statements?

It's OK the way it is now ..

The print statements are in the `def fetch_traits(..):` function / this function is not called in the server context / we need this print statements when running a `make data.traits`.

Reply:
Nice. I thought it was getting called everytime.